src/plugins/blocksmith/components/blocksmith/error-boundary/index.jsx

```diff
@@ -79,92 +99,114 @@ class ErrorBoundary extends React.Component {
-        <hr />
-        {blocks.map(({ block }) => {
-          if (block instanceof ContentBlock) {
-            let message = 'block';
-            const blockData = block.getData();
-            if (blockData && blockData.has('canvasBlock')) {
-              message = blockData.get('canvasBlock').schema().getCurie().getMessage();
-            }
-            return (
-              <div className="preview-component preview-component__error" key={block.getKey()}>
-                <p><strong>{`invalid ${message}`}</strong></p>
-                <p>{block.toString()}</p>
-              </div>
-            );
-          }
-          const message = block.schema().getCurie().getMessage();
-          let Component;
-          switch (message) {
-            case 'text-block':
-              // eslint-disable-next-line react/no-danger
-              Component = () => <div dangerouslySetInnerHTML={{ __html: block.get('text') }} />;
-              break;
-            case 'image-block':
-              Component = () => (
-                <>
-                  <p>{message}</p>
-                  <img src={damUrl(block.get('node_ref'), 'o', 'xs')} alt="thumbnail" />
-                </>
-              );
-              break;
-            case 'document-block':
-              Component = () => (
-                <>
-                  <p>{message}</p>
-                  <img src={damUrl(block.get('image_ref'), 'o', 'xs')} alt="thumbnail" />
-                </>
-              );
-              break;
-            case 'gallery-block':
-              Component = () => (
-                <>
-                  <p>{message}</p>
-                  <img src={damUrl(block.get('poster_image_ref', getNode(block.get('node_ref')).get('image_ref')), 'o', 'xs')} alt="thumbnail" />
-                </>
-              );
-              break;
-            case 'video-block':
-              Component = () => (
-                <>
-                  <p>{message}</p>
-                  <img src={damUrl(block.get('poster_image_ref', getNode(block.get('node_ref')).get('image_ref')), 'o', 'xs')} alt="thumbnail" />
-                </>
+      <Card>
+        <CardHeader>
+          Content
+          <kbd
+            className="text-dark bg-white text-uppercase"
+            style={{ fontFamily: 'inherit', fontSize: '15px' }}
+          >
+            <span style={{ fontSize: '11px' }}>Word Count</span>
+            <Badge color="light" className="ml-1 font-weight-normal" style={{ fontSize: '11.25px' }}>
+              {getWordCount(editorState)}
+            </Badge>
+          </kbd>
+        </CardHeader>
+        <CardBody indent>
+          <div className="blocksmith-error-boundary">
+            <p className="warning">{warning}</p>
+            <Button
+              size="md"
+              onClick={() => {
+                const result = validateBlocks(editorState);
+                this.handleRestoreBlocksmith({ ...result, editorState: result.validEditorState });
```

Comment:
@hunyoboy since you removed

```
    if (isValid && errorCount < MAX_ERROR_COUNT) {
      this.restoreBlocksmith(validEditorState, false);
      return null;
    }
```

this would need to pass isValid in, eh?

that being said - is all this change really necessary? i think just wrapping the entire component in the error boundary (which is a great idea btw) and the Raven changes is sufficient

Reply:
@efjacobson it'll still work but the main reason why i'm modifying it here since i believe it's not advisable or we should avoid if possible to put any side effects inside `render` function. also react [doc](https://reactjs.org/docs/react-component.html#render) here states clearly: 

> The render() function should be pure, meaning that it does not modify component state, it returns the same result each time it’s invoked, and it does not directly interact with the browser.

If you need to interact with the browser, perform your work in componentDidMount() or the other lifecycle methods instead. Keeping render() pure makes components easier to think about.

and on their error boundary [demo](https://codepen.io/gaearon/pen/wqvxGa?editors=0010). 

also there's a good discussion [here](https://stackoverflow.com/questions/48226268/calling-setstate-in-react-from-render-method/52024407#:~:text=Calling%20setState()%20inside%20render,state%20and%2For%20props%20directly) about it.

Reply:
@hunyoboy why does `handleRestoreBlocksmith` even have the `isValid` parameter? it is never provided

this approach seems less clear, im trying to understand the real benefit. i really think just moving the error-boundary and the Raven change is the win here

Reply:
@hunyoboy why does `handleRestoreBlocksmith` even have the `isValid` parameter? it is never provided

this approach seems less clear, im trying to understand the real benefit. i really think just moving the error-boundary and the Raven change is the win here

Reply:
@hunyoboy why does `handleRestoreBlocksmith` even have the `isValid` parameter? it is never provided

this approach seems less clear, im trying to understand the real benefit. i really think just moving the error-boundary and the Raven change is the win here

Reply:
@hunyoboy also, the way it is now is not going to create an infinite loop as discussed in that stackoverflow thread

Reply:
> @hunyoboy why does `handleRestoreBlocksmith` even have the `isValid` parameter? it is never provided
> 

@efjacobson yes it is because we are now setting that within the state. and yes we need to pass it now and it's kind of a drawback but not really since we're just computing this exactly as how many times we needed it regardless how many times the component is re-rerendered. current implementation is we're pulling it from `validateBlocks` w/c is inside render. and this render can be called many, many times. if you intentionally throw an exception and render an error-boundary, you will see that the component will still be re-rendered multiple times and will unnecessarily invoke 'validateBlocks` ( ex. raven can re-render component ). i don't think you'll agree with that. whereas if we extract it outside render, we do  not make those unnecessary calls.

Reply:
@hunyoboy 

> yes it is because we are now setting that within the state. and yes we need to pass it now and it's kind of a drawback but not really since we're just computing this exactly as how many times we needed it regardless how many times the component is re-rerendered
where is `isValid` passed into `handleRestoreBlocksmith`?

> if you intentionally throw an exception and render an error-boundary, you will see that the component will still be re-rendered multiple times and will unnecessarily invoke 'validateBlocks`
does it matter? let me ask you this - what problem are you trying to solve? is the additional complexity worth it to prevent extra renders in this edge case scenario? now this component has a lot more state that is just unnecessary - it can be derived from props..it is duplicate state and that should be avoided at all costs